README.md
TraCINg-Server
==============

A webserver gathering malware incidents and visualizing them in multiple ways.

TraCINg (TUD Cyber Incident moNitor with TUD as an abbreviation of Technische Universität Darmstadt)
is a project proposed by Emmanouil Vasilomanolakis from [CASED](http://www.cased.de/) (Center for
Advanced Security Research Darmstadt) visualizing attacks of malware on the internet.
Attacks are observed using honeypots especially the honeypot [dionaea](http://dionaea.carnivore.it/)
in conjunction with [jsonfeeds](https://github.com/Cyber-Incident-Monitor/jsonfeeds) and
the honeypot [HosTaGe](https://github.com/mip-it/hostage) but can be extended to use arbitrary
honeypots, intrusion detection systems (IDS) and similar software.

This product includes GeoLite data created by MaxMind, available from http://maxmind.com/.  
This project was inspired by but is not based on the [Honeynet Project](http://map.honeynet.org/).

## Features ##
### Backend ###
The backend consists internally of two servers: A HTTPS server receiving sensor data
and a HTTP server serving a website to visualize this data.
Sensors are honeypots (or intrusion detection systems) collecting information about attacks of malware.

The HTTP server acts like a simple webserver with static content. Dynamic content is served using
[Socket.IO](http://socket.io/).

The HTTPS servers purpose is to receive sensor data, to store them in the database and to broadcast it via
socket.io to every client currently viewing the content of the HTTP server.
The encryption is mandatory to protect the sensors identity by hiding the content of the data transmission.
This is necessary to avoid revealing the IP addresses of sensors by observing the transmitted content which
could lead to blacklisting of sensor IPs in malware.
Note that sensors are encouraged to hide their IP addresses using for example [Tor](https://www.torproject.org)
to avoid revealing their IPs in an observation of the traffic to the HTTPS server with the assumption that every sent
message to the HTTPS server is most likely sent from a sensor.

In order to receive genuine sensor data an authentification is used to recognize trustworthy sensors.
This authentification is based on a public-key infrastructure (PKI) using a certificate authority (CA) to
sign the client certificates.
The authentification is based on sending the client certificate at the TLS handshake and verifying it on the
server using the CA certificate.

Thus authorizing a sensor requires the sensor to send a certificate request to the CA and to get a valid
certificate from the CA.

### Frontend ###
The website visualizes malware attacks from the internet in five different ways (described below). Incidents
will be shown live if in **Live-View** or can be retrieved by a database query if in **Database-View**.
Additional features are an about and help screen to guide and inform the user.

#### 2D Country View ####
The country view shows a map which contains only country borders.
It can be moved and zoomed both with the mouse or the keyboard.  
Attacks are shown as a marker in the country where the attackers IP was mapped to. Hovering a marker shows
information about this specific attack and how many attacks originated from the same place.  
Additionally countries are colored in a red shade depending on the ratio of markers in that country.

#### 2D Map View ####
The map view behaves much like the country view but omits coloring of the countries. Instead it shows a
more detailed map using [OpenStreetMap](http://www.openstreetmap.org/) map material.

#### 3D Globe ####
The globe behaves much like the country view in the 3D space with the enhancement of adding a heatmap-like
view of the markers which can be toggled with the keyboard.

#### Table View ####
The table view shows a sort- and searchable table containing detailed information about each attack. If the sensor
was able to copy the malware the malwares md5sum is given in addition to a link to
[VirusTotal](https://www.virustotal.com/) where one can get more detailed information about this specific malware.
It is also possible to have a look into logs which dionaea creates to record every attack.

#### Statistics ####
The statistic shows either the number of attacks per country or per type in a specific time span.
The data can be filtered in several ways (the same is possible in **Database-View**):
* show only authorized sensors data
* select countries
* select attack types
* select sensor types
Note that statistics can only be applied to data queried from the database (though the statistic itself is independent
from **Live-View** and **Database-View**).

## Requirements ##
### System Packages ###
In order to run the server one must install the following packages (preferably with
the systems package management system):
* [openssl](http://www.openssl.org/)
* [sqlite3](http://www.sqlite.org/)
* [node.js](http://nodejs.org/)

For example using [pacman](https://wiki.archlinux.org/index.php/pacman): `pacman -S openssl sqlite3 nodejs`

### Node Packages ###
Additionally the following [npm](https://npmjs.org/) packages are required to be installed:
* [socket.io](https://npmjs.org/package/socket.io)
* [sqlite3](https://npmjs.org/package/sqlite3)
* [orm](https://npmjs.org/package/orm)
* [node-static](https://npmjs.org/package/node-static)
* [geoip](https://npmjs.org/package/geoip)
* [validator](https://npmjs.org/package/validator)

You can call `npm install` in the root directory of this repository to install the dependencies.

### Website Libraries ###
To run the website one must provide several external libraries (at least the javascript and css files)
in the **frontend/extern** folder:
* [bootstrap](http://getbootstrap.com/2.3.2/) (including images)
* [jVecorMap](http://jvectormap.com/), [world map](http://jvectormap.com/maps/world/world/)
* [leaflet](http://leafletjs.com/) (including images)
* [globe.js](https://github.com/Cyber-Incident-Monitor/globe.js) (including images), requires [three.js](http://threejs.org/)
* [highcharts](http://www.highcharts.com/)
* [jQuery](http://jquery.com/)
* [jQuery UI](http://jqueryui.com/) (including images)
* [jquery-ui-multiselect-widget](http://www.erichynds.com/blog/jquery-ui-multiselect-widget)
* [jrange](https://github.com/Cyber-Incident-Monitor/jrange)
* [jquery-throttle-debounce](http://benalman.com/projects/jquery-throttle-debounce-plugin/)
* [datatables](https://datatables.net/) (including images), [datatables bootstrap plugin](http://datatables.net/blog/Twitter_Bootstrap_2)

Instead of installing all these libraries manually we encourage you to use the provided
[fetch.sh](https://raw.github.com/Cyber-Incident-Monitor/TraCINg-Server/master/fetch.sh)
script to download them automatically along with the MaxMind GeoLiteCity database described in the next section.

### MaxMind GeoLiteCity Database ###
One must download the **GeoLiteCity.dat** file provided by MaxMind at
http://dev.maxmind.com/geoip/legacy/geolite/ and place it in the same folder as **index.js**.

### Certificates ###
To run the HTTPS server part one must provide at least a self signed server certificate
along with the corresponding private key.
To use the server to its full extent (with sensor authentification) one must prepare a
public-key infrastructure (PKI) containing a certificate authority (CA) which signs the
server and sensor certificates.
Hence one must provide the following files:
* server certificate (signed by the CA)
* server private key
* CA certificate

Note that the certificates and private keys must be provided in the pem format.

To test the functionality one may use the provided
[genKeyCert.sh](https://raw.github.com/Cyber-Incident-Monitor/TraCINg-Server/master/genKeyCert.sh)
script which generates CA, server, simulator and several client certificate/private key pairs in
the **ssl** folder. Note that these keys are weak (only 1024 bit long and not encrypted with a passphrase) and
are valid for just three days.

## Usage ##
To start the server execute `node index.js`. If the servers private key is encrypted you must unlock the
key by entering the passphrase.  
You may use the [simulator](https://github.com/Cyber-Incident-Monitor/TraCINg-Server/blob/master/simulator.py)
(requires python 3 along with the [Requests library](http://docs.python-requests.org/en/latest/)) to simulate
a sensor and thus test the functionality of the server.

### Configuration file ###
The server comes with a configuration file (config.json) which must be adapted to the users preferences:
```json
{
	"db": "sqlite:///tmp/test.sqlite?debug=true",
	"server": {
        	"httpPort": 8888,
        	"httpsPort": 9999,
        	"webroot": "./frontend"
	},
	"ssl":  {
		"keyPath": "ssl/server_key.pem",
		"certPath": "ssl/server_cert.pem",
		"caPath": "ssl/ca_cert.pem",
		"requestCert": true,
		"rejectUnauthorized": false
	}
}
```
* db: the path to the database storing attack data monitored by sensors
* geoip: the path to the GeoLiteCity.dat file provided by MaxMind
* server:
  * httpPort: the port of the HTTP server (serving the website)
  * httpsPort: the port of the HTTPS server (receiving sensor data)
  * webroot: the webroot containing the websites content
* ssl:
  * keyPath: the path to the server private key
  * certPath: the path to the server certificate
  * caPath: the path to the CA certificate
  * requestCert: if true the server requests a certificate to check sensors authenticity
  * rejectUnauthorized: if true unauthorized sensors are rejected

## Server Interface ##
A sensor must stick to the following JSON notation of a data entry to be able to send data to this server. Note that
the data entry must be in one line and must not be separated by line breaks as shown here for a better readability:
```json
{
	"sensor": {
		"name": "sensorName",
		"type": "sensorType"
	},
	"src": {
		"ip": "sourceIP",
		"port": "sourcePort"
	},
	"dst": {
		"ip": "destinationIP",
		"port": "destinationPort"
	},
	"type": "incidentTypeID",
	"log": "incidentLog",
	"md5sum": "incidentMd5sum",
	"date": "unixTimestamp"
}
```
This data must be sent via a POST message to the HTTPS server in order to be received correctly.
The sensor may send multiple datasets in one POST message each separated with a "\n".

The following table shows which data fields may be omitted and which are mandatory along with the default
values:

| Field		| Description			| Datatype	| Example		| Default Value			|
|---------------|-------------------------------|---------------|-----------------------|-------------------------------|
| sensor.name	| sensor name			| String	| Sensor1		| "Unknown"			|
| sensor.type	| sensor type			| String	| Honeypot		| "Unknown"			|
| src.ip	| attacker IP			| String	| 130.83.151.135	| **mandatory field**		|
| src.port	| attacker port			| Integer	| 54321			| 0				|
| dst.ip	| attacked IP (sensor IP)	| String	| 130.83.151.136	| *empty string*		|
| dst.port	| attacked port	(sensor port)	| Integer	| 80			| 0				|
| type		| attack type (cf. next table)	| Integer	| 11			| "Unknown"			|
| log		| attack log			| String	| TCP accept...		| *empty string*		|
| md5sum	| md5sum of a malware		| String	| 0e65972dce68...	| *empty string*		|
| date		| date of the attack		| Integer	| 1376645816		| *unix time of the server*	|

The following table shows the association between attack type numbers and attack type definitions:

| Attack Type 	| Attack Name 		| Attack Description												|
|---------------|-----------------------|---------------------------------------------------------------------------------------------------------------|
| 0		| Unknown		| The sensor could not determine the attack type								|
| 10		| Transport Layer	| The attacker connected to an open port, but did not interact with it						|
| 11		| Portscan		| The attacker tried to connect to a closed port								|
| 20		| Shellcode Injection	| The attacker successfully used an emulated security issue and would have been able to execute malicious code	|
| 30		| SQL			| Attack on a database server											|
| 31		| MySQL			| Attack on a MySQL database server										|
| 32		| MS SQL		| Attack on a Microsoft database server										|
| 40		| SMB			| Attack on a SMB file server											|
| 50		| VoIP			| Attack on a Voice over IP device										|

Note that these attack types are based on the ability of [dionaea](http://dionaea.carnivore.it/) to distinguish between
these types of attacks.

## Examples ##
Websites running this server:
* http://ssi.cased.de


Changelog.md
Changelog
=========

## v1.0.2 ##

* filtering for 'Unknown' incident types returns now all incidents without a known type (issue #30)

## v1.0.1 ##

* fix link to virustotal in table view

## v1.0.0 ##

* initial release


simulator.py
# TraCINg-Server - Gathering and visualizing cyber incidents on the world
#
# Copyright 2013 Matthias Gazzari, Annemarie Mattmann, André Wolski
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import requests
import json
import random
import time
import sys
import argparse
import time
import datetime

# based on http://stackoverflow.com/questions/10218486/set-nested-dict-value-and-create-intermediate-keys
from collections import defaultdict
recursivedict = lambda: defaultdict(recursivedict)

# constants and defaults
sensorName = "Simulator"
sensorType = "Honeypot"
url = "https://localhost:9999"
cert = ("ssl/simulator/simulator_cert.pem", "ssl/simulator/simulator_key.pem")
incidentTypes = {
	0: "Unknown",
	10: "Transport Layer",
	11: "Portscan",
	20: "Shellcode Injection",
	30: "SQL",
	31: "MySQL",
	32: "MS SQL",
	40: "SMB",
	50: "VoIP",
	60: "Invalid"		# invalid test case
}
predefined = {
	"sensor": {
		"name": "Predefined Sensor Name",
		"type": "Predefined Sensor Type",
	},
	"src": {
		"ip": "130.83.58.211",
		"port": 80,
	},
	"dst": {
		"ip": "192.30.252.130",
		"port": 22,
	},
	"type": 0,
	"log": "Predefined Log",
	"md5sum": "7867de13bf22a7f3e3559044053e33e7",
	"date": 1,
}

# Return a field only containing the mandatory field
def getMandatoryOnlyEntry():
	payload = recursivedict()
	payload["src"]["ip"] = getRandomIP()
	return payload

# Return a randomly generated entry using every fields possible
def getFullEntry(chooseRandomly):
	return {
		"sensor": {
			"name": sensorName,
			"type": sensorType,
		},
		"src": {
			"ip": getRandomIP(),
			"port": getRandomPort(),
		},
		"dst": {
			"ip": getRandomIP(),
			"port": getRandomPort(),
		},
		"type": getRandomIncident(),
		"log": getRandomLog(),
		"md5sum": getRandomMd5sum(),
		"date": getTime(chooseRandomly),
	}

# Return a randomly generated entry (mandatory field always set)
def getRandomizedEntry(chooseRandomly):
	payload = recursivedict()
	setRandomly(payload, "sensor", "name", sensorName)
	setRandomly(payload, "sensor", "type", sensorType)
	payload["src"]["ip"] = getRandomIP()
	setRandomly(payload, "src", "port", getRandomPort())
	setRandomly(payload, "dst", "ip", getRandomIP())
	setRandomly(payload, "dst", "port", getRandomPort())
	setRandomly(payload, "type", None, getRandomIncident())
	setRandomly(payload, "log", None, getRandomLog())
	setRandomly(payload, "md5sum", None, getRandomMd5sum())
	setRandomly(payload, "date", None, getTime(chooseRandomly))
	return payload

# Return a randomly set entry to be submitted
def getRandomlyEntry(mode, chooseRandomly):
	if mode == None:
		mode = random.randint(0, 2)
	# only set mandatory fields (source IP address)
	if mode == 0:
		payload = getMandatoryOnlyEntry()
	# set every possible field
	elif mode == 1:
		payload = getFullEntry(chooseRandomly)
	# set randomly fields (but always the mandatory field)
	else:
		payload = getRandomizedEntry(chooseRandomly)
	return payload

# Return either the current or a random time (32bit unix time)
def getTime(chooseRandomly):
	if chooseRandomly:
		return random.randint(0, 2**31 - 1)
	else:
		return int(time.time())

# Set a value with a probability of 0.5
def setRandomly(target, key1, key2, content):
	if random.randint(0, 1) == 0:
		if key2 == None:
			target[key1] = content
		else:
			target[key1][key2] = content

# Return a random port ranging from 0 to 65535
def getRandomPort():
	return random.randint(0, 2**16 - 1)

# Return a random IP address
def getRandomIP():
	ip = [random.randint(0, 255) for _ in range(4)]
	return '.'.join([str(e) for e in ip])

# Return a random incident chosen from incidentTypes dictionary
def getRandomIncident():
	return random.choice(list(incidentTypes.keys()))

# Return a random md5sum
def getRandomMd5sum():
	return str(hex(random.randint(0, 2**128 - 1)).lstrip("0x"))

# Return a random log message
def getRandomLog():
	return "Testlog with random value: " + str(random.randint(0, 2**10)) + "\nand a new line and <b>html</b>"

# Post the payload in json format to the server
def post(payload_json, url, cert, useCert, verify):
	try:
		if useCert:
			r = requests.post(url, cert=cert, verify=verify, data=payload_json)
		else:
			r = requests.post(url, verify=verify, data=payload_json)
		return payload_json + "\n --> " + str(r) + "\n>---\n" + r.text + "\n>---\n"
	except requests.exceptions.SSLError as e:
		print(e)
	except IOError as e:
		print("Either the cert, the key or both of them are not found.")
	except:
		 print("Unexpected error:", sys.exc_info()[0])
		 raise
	return ""
	
# define a positive integer type
def positiveInt(string):
	value = int(string)
	if value < 0:
		msg = string + " is not a positive integer"
		raise argparse.ArgumentTypeError(msg)
	return value

def main():
	# define arguments and their behaviour
	parser = argparse.ArgumentParser(description = "Simulate incident reports (single, multiple or in logs) and send them via HTTPS POST to a designated server.")
	parser.add_argument("-s", "--seed", help = "set the SEED to initialize the pseudorandom number generator", type = positiveInt)
	parser.add_argument("-m", "--mode", help = "determine which fields are sent (0: only mandatory fields, 1: every possible field, 2: at least the mandatory fields)", choices = [0, 1, 2], type = int)
	parser.add_argument("-n", "--number", help = "set the NUMBER of incident reports (or logs) to be sent (by default 1)", type = positiveInt, default = 1)
	parser.add_argument("-i", "--interval", help = "set the MIN and MAX time in ms between single incident reports (by default 1000 and 1000)", nargs = 2, metavar = ("MIN", "MAX"), type = positiveInt, default = [1000, 1000])
	parser.add_argument("-v", "--verbose", help = "show more details while sending incident reports", action = "store_true")
	parser.add_argument("-u", "--url", help = "set the server URL to sent the incident report(s) (by default " + url + ")", default = url)
	certFormated = str(cert).replace("(", "").replace(",", " and")
	parser.add_argument("-c", "--cert", help = "set the CERT and private KEY path to be used (by default " + certFormated, nargs = 2, metavar = ("CERT", "KEY"), default = cert)
	parser.add_argument("-nc", "--no-cert", help = "disable certificate usage", action = "store_true")
	parser.add_argument("-vc", "--verify-cert", help = "disable server certificate verification", action = "store_true")
	parser.add_argument("-lf", "--log-format", help = "send multiple incident reports in one log (by default 3 to 10 reports per log)", action = "store_true")
	parser.add_argument("-ls", "--log-size", help = "set the MIN and MAX number of incident reports per log (by default MIN = 3 and MAX = 10)", nargs = 2, metavar = ("MIN", "MAX"), type = positiveInt, default = [3, 10])
	parser.add_argument("-r", "--random-time", help = "set the timestamp at random instead of using the current time", action = "store_true")
	group = parser.add_mutually_exclusive_group()
	group.add_argument("-p", "--predefined", help = "send a predefined incident report (cf. the source of this program for details about the predefined report)", action = "store_true")
	group.add_argument("-cu", "--custom", help = "apply a custom incident REPORT in JSON format, for example: '{\"src\":{\"ip\":\"192.30.252.130\"}}' (put the incident report into single quotes to prevent the shell from removing the double quotes)", metavar = "REPORT", type = json.loads)
	args = parser.parse_args()
	
	# init seed
	if args.seed:
		seed = args.seed
	else:
		seed = random.randint(0, sys.maxsize)
	random.seed(seed)
	
	# send incidents
	if args.number > 0:
		for i in range(0, args.number):
			# send multiple entries in one log separated by \n
			if args.custom is not None:
				entry = args.custom
			elif args.predefined:
				entry = predefined
			elif args.log_format:
				entry = ""
				size = random.randint(args.log_size[0], args.log_size[1])
				for j in range(0, size):
					entry += getRandomlyEntry(args.mode, args.random_time) + "\n"
			else:
				# send a single entry
				entry = getRandomlyEntry(args.mode, args.random_time)
			# post the entry
			result = post(json.dumps(entry), args.url, args.cert, not args.no_cert, args.verify_cert)
			# print server reply
			if args.verbose:
				print("-----------------------------------------------")
				print("Attack No.:", i + 1, "of", args.number)
				try:
					print("Date (UTC):", datetime.datetime.utcfromtimestamp(entry["date"]))
				except:
					print("Date is not supplied")
				try:
					print("Incident type:", incidentTypes[entry["type"]])
				except:
					print("Incident type is not supplied")
				print("Server response:\n")
				print(result)
			else:
				print("Attack No.:", i + 1, "of", args.number)
			# avoid sleep in the last loop
			if i < args.number - 1:
				time.sleep(random.randint(args.interval[0], args.interval[1])/1000)
	
	if args.verbose:
		print("-----------------------------------------------")
	
	# determine new args applying the seed if missing to the previous args
	newArgs = " ".join(sys.argv[1:])
	if not args.seed:
		newArgs += " --seed " + str(seed)
	
	# add single quotes to custom json entry argument
	# (prevent the shell removing the required double quotes)
	if args.custom is not None:
		trimmedEntry = str(entry).replace(" ", "").replace("'", '"')
		newArgs = newArgs.replace(trimmedEntry, "'" + trimmedEntry + "'")
	
	# print used arguments with an additional example command
	print("To reproduce this simulation use the following argument(s): ")
	print("\t" + newArgs)
	print("For example using the following command: ")
	print("\t python3 " + sys.argv[0], newArgs)

if __name__ == '__main__':
	main()
